internal/lockclient/simple_client.go

```diff
@@ -20,114 +26,306 @@ type SimpleClient struct {
+
+	// sessions holds the mapping of a process to a session.
+	sessions map[id.ID]session.Session
+	// sessionTimers maintains the timers for each session,
+	sessionTimers map[id.ID]chan struct{}
+	// sessionAcquisitions has a list of all the acquisitions
+	// from a particular process. This has no knowledge of
+	// whether the process owning the lock has an active session
+	// or not, this guarantee has to be ensured by the client.
+	sessionAcquisitions map[id.ID][]lockservice.Descriptors
 }
 
 // NewSimpleClient returns a new SimpleClient of the given parameters.
 // This client works with or without the existance of a cache.
-func NewSimpleClient(config *lockservice.SimpleConfig, cache *cache.LRUCache) *SimpleClient {
+func NewSimpleClient(config *lockservice.SimpleConfig, log zerolog.Logger, cache *cache.LRUCache) *SimpleClient {
+	clientID := id.Create()
+	sessions := make(map[id.ID]session.Session)
+	sessionTimers := make(map[id.ID]chan struct{})
+	sessionAcquisitions := make(map[id.ID][]lockservice.Descriptors)
 	return &SimpleClient{
-		config: config,
-		cache:  cache,
+		config:              config,
+		cache:               cache,
+		id:                  clientID,
+		log:                 log,
+		sessions:            sessions,
+		sessionTimers:       sessionTimers,
+		sessionAcquisitions: sessionAcquisitions,
 	}
 }
 
 var _ Client = (*SimpleClient)(nil)
 
+// Connect lets the user process to establish a connection with the
+// client.
+func (sc *SimpleClient) Connect() session.Session {
+	sessionID := id.Create()
+	processID := id.Create()
+	session := session.NewSession(sessionID, sc.id, processID)
+	sc.sessions[processID] = session
+	sc.startSession(processID)
+	sc.log.
+		Debug().
+		Str(processID.String(), "connected").
+		Msg("session created")
+	return session
+}
+
 // Acquire allows the user process to acquire a lock.
-func (sc *SimpleClient) Acquire(d lockservice.Descriptors) error {
-	return sc.acquire(d)
+// This returns a "session expired" error if the session expires when
+// the lock is being acquired.
+//
+// All locks acquired during the session will be revoked if the session
+// expires.
+func (sc *SimpleClient) Acquire(d lockservice.Object, s session.Session) error {
+	sc.mu.Lock()
+	if _, ok := sc.sessions[s.ProcessID()]; !ok {
+		sc.mu.Unlock()
+		return ErrSessionInexistent
+	}
+	sc.mu.Unlock()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	close := make(chan struct{}, 1)
+	go func() {
+		for {
+			sc.mu.Lock()
+			select {
+			case <-sc.sessionTimers[s.ProcessID()]:
+				cancel()
+				sc.log.
+					Debug().
+					Str(s.ProcessID().String(), "user process").
+					Msg("session expired, starting graceful shutdown")
+				sc.mu.Unlock()
+				sc.gracefulSessionShutDown(s.ProcessID())
+				return
+			case <-close:
+				sc.mu.Unlock()
+				return
+			default:
+				sc.mu.Unlock()
+			}
+		}
+	}()
+	ld := lockservice.NewLockDescriptor(d.ID(), s.ProcessID().String())
+	err := sc.acquire(ctx, ld)
```

Comment:
shouldn't there be a lock() and unlock() before and after acquire()?

Reply:
ohhhhh nice, you've handled the timeout in acquire() :D

Reply:
So this is a principle Ive used generally while writing raft w.r.t locks - if a function needs something to be locked, let it lock it inside itself and dont lock it before its called. It keeps the locks cleaner and within the scope of the problem.